src/toolkits/drawing_classifier/drawing_classifier.cpp

```diff
@@ -345,8 +348,9 @@ void drawing_classifier::init_training(
   }
 
   // Report to the user what GPU(s) is being used.
-  std::vector<std::string> gpu_names = training_compute_context_->gpu_names();
-  print_training_device(gpu_names);
+  /* The following two lines segfault!!! */
+  // std::vector<std::string> gpu_names = training_compute_context_->gpu_names();
+  // print_training_device(gpu_names);
```

Comment:
We should discuss whether this change should go in or not. Locally, I get a Segmentation Fault, but commenting this line out gets rid of it.

Reply:
We should figure out this one. I don't get this segmentation fault.

Reply:
What is the backtrace? Which compute context is affected?

Reply:
@nickjong `tf_compute_context`.

Backtrace as below:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0xa0)
    frame #0: 0x000000010709285c Python`list_dealloc + 28
Python`list_dealloc:
->  0x10709285c <+28>: movl   0xa0(%r14), %eax
    0x107092863 <+35>: cmpl   $0x31, %eax
    0x107092866 <+38>: jg     0x107092880               ; <+64>
    0x107092868 <+40>: incl   %eax
Target 0: (python) stopped.
(lldb) up
frame #1: 0x0000000149320816 libtctensorflow.so`pybind11::handle::dec_ref(this=0x00007ffee8ba35e0) const & at pytypes.h:197:39
   194             preferable to use the `object` class which derives from `handle` and calls
   195             this function automatically. Returns a reference to itself.
   196         \endrst */
-> 197         const handle& dec_ref() const & { Py_XDECREF(m_ptr); return *this; }
   198     
   199         /** \rst
   200             Attempt to cast the Python object into the given C++ type. A `cast_error`
(lldb) up
frame #2: 0x0000000149320765 libtctensorflow.so`pybind11::object::~object(this=0x00007ffee8ba35e0) at pytypes.h:240:17
   237         /// Move constructor; steals the object from ``other`` and preserves its reference count
   238         object(object &&other) noexcept { m_ptr = other.m_ptr; other.m_ptr = nullptr; }
   239         /// Destructor; automatically calls `handle::dec_ref()`
-> 240         ~object() { dec_ref(); }
   241     
   242         /** \rst
   243             Resets the internal pointer to ``nullptr`` without without decreasing the
(lldb) up
frame #3: 0x000000014931eba5 libtctensorflow.so`pybind11::object::~object(this=0x00007ffee8ba35e0) at pytypes.h:240:15
   237         /// Move constructor; steals the object from ``other`` and preserves its reference count
   238         object(object &&other) noexcept { m_ptr = other.m_ptr; other.m_ptr = nullptr; }
   239         /// Destructor; automatically calls `handle::dec_ref()`
-> 240         ~object() { dec_ref(); }
   241     
   242         /** \rst
   243             Resets the internal pointer to ``nullptr`` without without decreasing the
(lldb) up
frame #4: 0x000000014931f227 libtctensorflow.so`turi::neural_net::tf_compute_context::gpu_names(this=0x00007fbe7504a2a0) const at tf_compute_context.cpp:283:1
   280       });
   281     
   282       return gpu_device_names;
-> 283     }
   284     
   285     
   286     std::unique_ptr<model_backend> tf_compute_context::create_object_detector(
```

Reply:
What if we move the `gpu_devices` local variable here
https://github.com/apple/turicreate/blob/25eab55fd492f355f8667ddd2cc60e26697d8681/src/ml/neural_net/tf_compute_context.cpp#L270
inside the lambda?

I wonder if we can't release the GIL before destroying it on function return